src/types/presentation.ts

```diff
@@ -0,0 +1,110 @@
+import type { IIdentityService } from '../services/identity/index.js';
+import type { CustomerEntity } from '../database/entities/customer.entity.js';
+import { CheqdW3CVerifiableCredential } from './credential.js';
+import { CommonReturn } from './shared.js';
+import { JWT_PROOF_TYPE } from './constants.js';
+
+export interface ICheqdPresentation extends UnsignedPresentation {
+	proof: {
+		type: string;
+		jwt: string;
+	};
+}
+
+export interface IJWTPayloadVP extends JwtPayload {
+	vp: ICheqdPresentation;
+}
+
+// ToDo: make unit tests
+export class CheqdW3CVerifiablePresentation extends CommonReturn implements ICheqdPresentation {
+	holder: string;
+	verifiableCredential?: CheqdW3CVerifiableCredential[];
+	type?: string[] | string;
+	'@context': ContextType;
+	verifier?: string[];
+	issuanceDate?: string;
+	expirationDate?: string;
+	id?: string;
+	proof: {
+		type: string;
+		jwt: string;
+	};
+
+	constructor(w3Presentation: W3CVerifiablePresentation) {
+		super();
+		let presentation: VerifiablePresentation;
+		presentation = w3Presentation as VerifiablePresentation;
+		if (typeof w3Presentation === 'string') {
+			presentation = this.fromVPCompactJWT(w3Presentation);
+		}
+
+		this['@context'] = presentation['@context'];
+		this.type = presentation.type;
+		this.verifiableCredential = presentation.verifiableCredential?.map(
+			(vc) => new CheqdW3CVerifiableCredential(vc)
+		);
+		this.holder = presentation.holder;
+		this.verifier = presentation.verifier;
+        this.issuanceDate = presentation.issuanceDate;
+		this.proof = (presentation as ICheqdPresentation).proof;
+	}
+
+	public fromVPCompactJWT(jwt: CompactJWT): VerifiablePresentation {
+		const decoded = jwtDecode(jwt) as IJWTPayloadVP;
+
+		if (!Object.keys(decoded).includes('vp')) {
+			throw new Error('JWT does not contain a verifiable presentation');
+		}
+		const presentation: W3CVerifiablePresentation = decoded.vp;
+		// As described here: https://www.w3.org/TR/vc-data-model/#jwt-encoding
+		// iss - holder
+		presentation.holder = decoded.iss as string;
+		if (typeof decoded.aud === 'string') {
+			presentation.verifier = [decoded.aud] as string[];
+		} else if (Array.isArray(decoded.aud)) {
+			presentation.verifier = decoded.aud as string[];
+		}
+        // nbf - issuanceDate
+        presentation.issuanceDate = new Date((decoded.nbf as number) * 1000).toISOString();
+		presentation.proof = {
+			type: JWT_PROOF_TYPE,
+			jwt: jwt,
+		};
+		return presentation as VerifiablePresentation;
+	}
+
+	public async makeFeePayment(agent: IIdentityService, customer: CustomerEntity): Promise<ICommonErrorResponse> {
+		if (!this.verifiableCredential) {
+			return this.returnError(
+				StatusCodes.BAD_REQUEST,
+				'Verifiable credentials are not placed in presentation. Cannot make fee payment.'
+			);
+		}
+
+		const feePaymentResults = await Promise.all(
+			this.verifiableCredential.map(async (credential) => {
+				return await credential.makeFeePayment(agent, customer);
```

Comment:
This assumes that all the credentials have an encrypted status list , Is this expected behaviour in did-provider-cheqd?

In a presentation I think we should allow both encrypted and unencrypted one's and we make payment only to the one's which are encrypted. we can filter out the one's which are encrypted and run this loop only for them.

Or if we think all credentials should be encrypted, we should run a pre-check before making feePayment in a loop. By that way we can minimize cases where a payment is made and we still return an error

Reply:
Yup, we have such check inside `makeFeePayment` function for the credential class. Cause anyway we need to ask for the statusList I moved such check inside the makeFeePayment function. If the statusList is not `encrypted` it just returns `Ok`

Reply:
Yup, we have such check inside makeFeePayment function for the credential class. Cause anyway we need to ask for the statusList I moved such check inside the makeFeePayment function. If the statusList is not encrypted it just returns Ok

Reply:
I see unecrypted is handled, but an error is thrown when a credential has no credential status, even that can be included in the presenation. Either we filter that beforehand before calling makeFeePayment or returnOk inside makeFeePayment

Reply:
@anikitinDSR I see it's handled for unencrypted statuslist, but when one of the credential doesn't have a verification status, it is failing right? 

We should filter them beforehand and call makeFeePayment only for credentials with statusList, or returnOk inside makeFeePayment.